src/blog/first-100k-lines.vmd

```diff
@@ -0,0 +1,415 @@
+---
+title: How to Survive Your First 100,000 Lines
+author: Evan Ovadia
+date: Apr 9, 2022
+realm: blog
+path: blog/first-100k-lines
+layout: annotated
+namespace: c-blog m-annotated
+sponsor: us
+---
+
+
+
+The Vale compiler just hit its [100,000th line of code.](linkhere)
+
+
+This is an article about how we kept it from *collapsing under its own weight and exploding,* as many projects do.
+
+
+Ironically, most of these software engineering techniques didn't come from my time at Google, they instead came from my work on the Vale compiler and game development. [# Though I did later apply a lot of these architecture techniques to [Google Earth](https://earth.google.com/web/).] So, some of these might be surprising to my engineer comrades out there.
+
+
+These techniques range from determinism, to testing, to type-system techniques, to general architectural best-practices.
+
+
+And since I'm a language geek, I'll throw in some nonsense about languages!
+
+
+# If you think you're using enough assertions, you're wrong
+
+I'm not exaggerating when I say assertions are the greatest thing since sliced arrays, and have cut my debugging time by half.
+
+
+<div style="text-align: center"><a href="https://i.redd.it/e1ck7vngll811.png"><img src="/images/squirrel.jpg" width="60%"/></a></div><div style="opacity: .8; text-align: center"><a href="https://i.redd.it/e1ck7vngll811.png">In your darkest hour, the assertion arrives.</a></div>
+
+
+For those unfamiliar, an assertion is basically an if-statement with a panic:
+
+```
+if thing_that_i_assume_is_four != 4 {
+  // Prints "Thing isn't 4!" and halts the program.
+  panic("Thing isn't 4!");
+}
+```
+
+and a lot of languages have an `assert` function, so it can be written like:
+
+```
+assert(thing_that_i_assume_is_four == 4, "Thing isn't 4!");
+```
+
+
+
+Rule of thumb: Every time you _think_ something is true about your data, and it's not ensured by the type system, add an assertion that checks it.
+
+Don't just occasionally use them, _drench_ your code in them. When your code so much as sneezes, they should shake off like sand after the beach. _This is the way._
+
+
+We once had a bug in the post-parser which caused some data corruption, and _thank goodness_ it was caught by an assertion in the final stage of the compiler. [# Previously known as the Hammer stage, though now it has the more boring name of FinalAstSimplifier.] If that assertion didn't catch it, it would have made it into the LLVM-generated code, and would have taken _hours_ to track down.
+
+
+Such miracles are why the Vale compiler has an entire *1,795 assertions.*
+
+
+<slice/>
+
+
+# Lean on the Type System
```

Comment:
Maybe link to an article in a note for this section, so readers can reference that since information in this section is a bit spare.